File: macos/ArcaneDispatchTunnel/Bonded/BondedScheduler.swift
// Per-packet link selector for the bonded transport (Speed mode v0).
//
// Mirror of `lib/bonded/bonded_scheduler.dart`. Keep the algorithm in sync;
// the Dart side has the unit tests that pin behaviour, so any deviation
// here will be caught by the loopback compatibility test once we add a
// cross-language test rig (Phase 8 deliverable).
//
// Algorithm: credit-based picker.
//
//   credit(link) = bandwidthBytesPerSec(link) × rtt(link) − inflightBytes(link)
//
// 1. Filter eligible links by priority bucket (primary → secondary → backup)
//    and health (drop unhealthy/disabled).
// 2. Compute credit for each link in the best bucket; pick the highest.
// 3. If multiple links tie within 1e-6, fall back to weighted RR so the
//    bond converges to the configured weights when credit signals are tied.
// 4. Book the bytes (`inflightBytes += chunkLen`) so subsequent picks see
//    realistic credit. Caller MUST call `completeSend(linkId, bytes)`
//    on ACK / drop so the counter unwinds.

import Foundation

/// Priority bucket mirrors `Link.priority` from the Dart side.
public enum BondedLinkPriority: String, Codable {
    case primary
    case secondary
    case backup
    case never
}

/// Health status mirrors `LinkStatus`.
public enum BondedLinkStatus: String, Codable {
    case healthy
    case degraded
    case unhealthy
    case disabled
    case unknown
}

/// Snapshot of the data the scheduler needs to pick a link. Decoupled
/// from `Link`/`LinkMetric` so the bonded layer doesn't pull in the
/// policy types directly.
public struct BondedLinkState {
    public let linkId: String
    public let wireId: UInt16
    public var priority: BondedLinkPriority
    public var status: BondedLinkStatus
    public var weight: Int
    public var rttMs: Double
    public var bandwidthBps: Double
    public var inflightBytes: Int
    /// Observed packet-loss fraction, 0.0–1.0. Drives Streaming-mode's
    /// duplicate-on-loss decision. 0 means "unknown" — strategies treat
    /// it as "no problem".
    public var lossFraction: Double

    public init(
        linkId: String,
        wireId: UInt16,
        priority: BondedLinkPriority = .primary,
        status: BondedLinkStatus = .unknown,
        weight: Int = 1,
        rttMs: Double = 50.0,
        bandwidthBps: Double = 1_000_000.0,
        inflightBytes: Int = 0,
        lossFraction: Double = 0.0
    ) {
        self.linkId = linkId
        self.wireId = wireId
        self.priority = priority
        self.status = status
        self.weight = weight
        self.rttMs = rttMs
        self.bandwidthBps = bandwidthBps
        self.inflightBytes = inflightBytes
        self.lossFraction = lossFraction
    }
}

/// Result of a single pick. The wireId is what the framer stamps into
/// the outgoing bonded frame.
public struct BondedSchedulingDecision {
    public let linkId: String
    public let wireId: UInt16
    public let credit: Double
    public let wasRoundRobinFallback: Bool
}

/// Speed-mode credit picker. Single-threaded — call from a serial queue.
public final class BondedScheduler {
    /// Insertion-ordered link state. We need stable iteration for the RR
    /// fallback, so `OrderedDictionary` would be nicer but Foundation only
    /// gives us `Dictionary` + a parallel `order` array.
    private var statesMap: [String: BondedLinkState] = [:]
    private var order: [String] = []
    private var rrCursor: Int = 0

    public let minBandwidthBps: Double
    public let minRttMs: Double

    public init(minBandwidthBps: Double = 50_000.0, minRttMs: Double = 1.0) {
        self.minBandwidthBps = minBandwidthBps
        self.minRttMs = minRttMs
    }

    public var linkCount: Int { statesMap.count }

    /// Read-only snapshot for callers that need to iterate. We copy so
    /// the consumer can't mutate scheduler internals.
    public var states: [String: BondedLinkState] { statesMap }

    /// Replace the link set. Preserves existing `inflightBytes` for links
    /// that survive the update; dropped links lose theirs (the caller
    /// should `completeSend` first if accounting matters).
    public func updateLinks(_ incoming: [BondedLinkState]) {
        var next: [String: BondedLinkState] = [:]
        var nextOrder: [String] = []
        for s in incoming {
            var copy = s
            if let prior = statesMap[s.linkId] {
                copy.inflightBytes = prior.inflightBytes
            }
            next[s.linkId] = copy
            nextOrder.append(s.linkId)
        }
        statesMap = next
        order = nextOrder
        if order.isEmpty {
            rrCursor = 0
        } else {
            rrCursor = rrCursor % order.count
        }
    }

    /// Pick a link for an outbound packet of `bytes` bytes. Returns `nil`
    /// when no link is eligible — caller should drop or queue.
    ///
    /// `inflightFraction` scales the BDP cap used in the credit formula.
    /// Speed mode passes 1.0 (default); Streaming mode shrinks it to 0.25
    /// to keep queueing delay minimal. Values outside [0.05, 1.0] are
    /// clamped.
    public func pickLink(bytes: Int, inflightFraction: Double = 1.0) -> BondedSchedulingDecision? {
        if statesMap.isEmpty || bytes <= 0 { return nil }
        var frac = inflightFraction
        if frac < 0.05 { frac = 0.05 }
        if frac > 1.0 { frac = 1.0 }
        let eligible = eligibleByPriority()
        if eligible.isEmpty { return nil }

        var bestCredit: Double = -.infinity
        var winner: String? = nil
        for id in eligible {
            let s = statesMap[id]!
            let c = credit(for: s, fraction: frac)
            if c > bestCredit {
                bestCredit = c
                winner = id
            }
        }

        var wasRR = false
        if winner == nil {
            // Defensive — only happens if every credit was -inf (e.g.
            // someone fed us NaN bandwidth). Pick the first eligible.
            winner = eligible.first
            wasRR = true
            bestCredit = 0
        } else {
            var ties = 0
            for id in eligible {
                let c = credit(for: statesMap[id]!, fraction: frac)
                if abs(bestCredit - c) < 1e-6 { ties += 1 }
            }
            if ties > 1 {
                winner = roundRobinPick(eligible: eligible)
                wasRR = true
            }
        }

        // Book inflight bytes.
        let winId = winner!
        var s = statesMap[winId]!
        s.inflightBytes += bytes
        statesMap[winId] = s
        return BondedSchedulingDecision(
            linkId: winId,
            wireId: s.wireId,
            credit: bestCredit,
            wasRoundRobinFallback: wasRR
        )
    }

    /// Manually book `bytes` of inflight on `linkId`. Strategies that
    /// synthesise their picks (e.g. Redundant duplicating to every link)
    /// call this directly so credit math stays accurate when the user
    /// later switches back to Speed mode.
    public func bookInflight(linkId: String, bytes: Int) {
        guard var s = statesMap[linkId] else { return }
        s.inflightBytes += bytes
        statesMap[linkId] = s
    }

    /// Release `bytes` from the link's inflight counter on ACK / timeout.
    public func completeSend(linkId: String, bytes: Int) {
        guard var s = statesMap[linkId] else { return }
        s.inflightBytes = max(0, s.inflightBytes - bytes)
        statesMap[linkId] = s
    }

    /// Public credit accessor for strategies (e.g. `LocalStrategy`) that
    /// want to evaluate non-default link populations without invoking the
    /// booking side-effect of `pickLink`. Returns 0 when the link is
    /// unknown. Mirrors Dart `BondedScheduler.creditFor`.
    public func creditFor(linkId: String, inflightFraction: Double = 1.0) -> Double {
        guard let s = statesMap[linkId] else { return 0 }
        var frac = inflightFraction
        if frac < 0.05 { frac = 0.05 }
        if frac > 1.0 { frac = 1.0 }
        return credit(for: s, fraction: frac)
    }

    /// Diagnostic alias retained for early tests. New code should call
    /// `creditFor(linkId:inflightFraction:)`.
    public func creditForTest(linkId: String, inflightFraction: Double = 1.0) -> Double {
        creditFor(linkId: linkId, inflightFraction: inflightFraction)
    }

    public func inflightForTest(linkId: String) -> Int {
        statesMap[linkId]?.inflightBytes ?? 0
    }

    // MARK: - internals

    private func credit(for s: BondedLinkState, fraction: Double) -> Double {
        let bw = s.bandwidthBps < minBandwidthBps ? minBandwidthBps : s.bandwidthBps
        let rttMs = s.rttMs < minRttMs ? minRttMs : s.rttMs
        let rttSec = rttMs / 1000.0
        let bdp = bw * rttSec * fraction
        return bdp - Double(s.inflightBytes)
    }

    /// Eligible links inside the best-available priority bucket. We don't
    /// mix buckets — that's the policy engine's job.
    private func eligibleByPriority() -> [String] {
        if statesMap.isEmpty { return [] }
        for bucket: BondedLinkPriority in [.primary, .secondary, .backup] {
            var ids: [String] = []
            for id in order {
                let s = statesMap[id]!
                if s.priority != bucket { continue }
                if s.status == .unhealthy || s.status == .disabled { continue }
                ids.append(id)
            }
            if !ids.isEmpty { return ids }
        }
        return []
    }

    private func roundRobinPick(eligible: [String]) -> String {
        if eligible.count == 1 {
            rrCursor = (rrCursor + 1) % max(order.count, 1)
            return eligible.first!
        }
        let start = rrCursor
        var picked = eligible.first!
        for i in 0..<order.count {
            let idx = (start + i) % order.count
            let candidate = order[idx]
            if eligible.contains(candidate) {
                picked = candidate
                rrCursor = (idx + 1) % order.count
                return picked
            }
        }
        return picked
    }
}

File: lib/bonded/bonded_scheduler.dart
// Per-packet link selector for the bonded transport (Speed mode v0).
//
// Mirrored on the Swift side by
// `macos/ArcaneDispatchTunnel/Bonded/BondedScheduler.swift`. The contract is
// "given a chunk of bytes, return the [linkId] that should carry it." The
// scheduler is *not* responsible for transmission — it only picks. The actual
// UDP send happens in the session driver so the scheduler stays
// trivially-testable.
//
// Algorithm (Speed mode):
//
//   credit(link) = bandwidthBytesPerSec(link) × rtt(link) − inflightBytes(link)
//
// On every `pickLink`:
//   1. Drop links whose status isn't healthy/degraded and whose priority is
//      `never`. Empty result → returns null (caller's job to apply policy).
//   2. Compute credit for each remaining link. If multiple links tie (e.g.
//      identical bandwidth/RTT, all idle), fall back to a stable weighted
//      round-robin so the load distribution converges to the configured
//      weights when credit signals are absent.
//   3. Once a link is picked, the scheduler "books" the bytes (adds to its
//      inflight counter) so the next pick reflects the in-progress send.
//      The caller MUST call [completeSend] once the bytes have been ACKed or
//      timed out, otherwise the link will look perpetually congested.
//
// Bandwidth comes from the latest [LinkMetric.throughputBytesPerSec]; if the
// metric is missing or stale we default to a small floor so the link still
// participates rather than starving.

import 'dart:math' as math;

import '../core/link.dart';
import '../core/link_metric.dart';

/// Snapshot of the data the scheduler needs to make a decision. Decoupled
/// from `LinkMetric` so the bonded layer doesn't have to import the probe
/// machinery; the controller adapts metrics into [BondedLinkState] before
/// pushing them in.
class BondedLinkState {
  /// Foreign key into `Link.id`.
  final String linkId;

  /// Numeric id stamped into the wire frame (`linkId` field). Picked by the
  /// session at registration time; we keep it separate from the string id so
  /// the protocol stays compact.
  final int wireId;

  /// Routing priority bucket from the policy.
  final LinkPriority priority;

  /// Health from the supervisor.
  final LinkStatus status;

  /// Integer weight inside the priority bucket. Drives the RR fallback.
  final int weight;

  /// Smoothed RTT estimate, milliseconds. Higher = worse credit.
  final double rttMs;

  /// Throughput estimate from the probe, bytes/sec. Higher = more credit.
  final double bandwidthBps;

  /// Bytes currently in flight on this link (booked by [BondedScheduler.pickLink]
  /// and decremented by [BondedScheduler.completeSend]).
  final int inflightBytes;

  /// Observed packet-loss fraction, 0.0–1.0. Strategies treat 0 as
  /// "unknown".
  final double lossFraction;

  const BondedLinkState({
    required this.linkId,
    required this.wireId,
    this.priority = LinkPriority.primary,
    this.status = LinkStatus.unknown,
    this.weight = 1,
    this.rttMs = 50.0,
    this.bandwidthBps = 1_000_000.0,
    this.inflightBytes = 0,
    this.lossFraction = 0.0,
  });

  BondedLinkState copyWith({
    LinkPriority? priority,
    LinkStatus? status,
    int? weight,
    double? rttMs,
    double? bandwidthBps,
    int? inflightBytes,
    double? lossFraction,
  }) {
    return BondedLinkState(
      linkId: linkId,
      wireId: wireId,
      priority: priority ?? this.priority,
      status: status ?? this.status,
      weight: weight ?? this.weight,
      rttMs: rttMs ?? this.rttMs,
      bandwidthBps: bandwidthBps ?? this.bandwidthBps,
      inflightBytes: inflightBytes ?? this.inflightBytes,
      lossFraction: lossFraction ?? this.lossFraction,
    );
  }

  /// Build a state from a [Link] + optional [LinkMetric]. Convenience for
  /// the policy-engine → scheduler glue layer.
  factory BondedLinkState.fromLink(
    Link link, {
    required int wireId,
    LinkMetric? metric,
  }) {
    // We use bpsOut (egress throughput) for the credit formula since the
    // scheduler is choosing outbound paths. Falls back to bpsIn, then to a
    // sensible default when neither is sampled yet so newly-added links
    // still participate in the first decision rather than starving for a
    // probe cycle.
    double bandwidth = metric?.bpsOut ?? metric?.bpsIn ?? 1_000_000.0;
    double rtt = metric?.rttMs ?? 50.0;
    double loss = metric?.loss ?? 0.0;
    return BondedLinkState(
      linkId: link.id,
      wireId: wireId,
      priority: link.priority,
      status: link.status,
      weight: link.weight,
      rttMs: rtt,
      bandwidthBps: bandwidth,
      lossFraction: loss,
    );
  }
}

/// Result of a single scheduling decision. The wireId is what the framer
/// stamps into the outgoing packet.
class BondedSchedulingDecision {
  /// Foreign key into the policy.
  final String linkId;

  /// Wire id stamped into the bonded frame.
  final int wireId;

  /// Credit at the moment of selection — exposed mostly for diagnostics /
  /// metrics. UI can plot "winning credit per pick" to debug starvation.
  final double credit;

  /// True iff this pick was made via RR fallback because credit signals
  /// tied. Useful for debugging "why didn't link X get traffic?".
  final bool wasRoundRobinFallback;

  const BondedSchedulingDecision({
    required this.linkId,
    required this.wireId,
    required this.credit,
    this.wasRoundRobinFallback = false,
  });
}

/// Speed-mode credit-based scheduler.
///
/// Lifecycle: build once per session, push state with [updateLinks] whenever
/// policy or metrics change, ask for a decision with [pickLink], settle the
/// inflight counter with [completeSend] once a packet is ACKed/dropped.
class BondedScheduler {
  final Map<String, BondedLinkState> _states = <String, BondedLinkState>{};
  // Stable iteration order for the RR fallback. Whenever updateLinks adds /
  // removes links we recompute the cursor's modulus.
  final List<String> _order = <String>[];
  int _rrCursor = 0;

  /// Floor for the bandwidth estimate. Without this an idle link with a
  /// brand-new metric of 0 B/s would always lose, even when it's the only
  /// healthy link.
  final double minBandwidthBps;

  /// Floor for the RTT estimate (ms). 1 ms keeps the credit formula from
  /// degenerating into infinity on loopback.
  final double minRttMs;

  BondedScheduler({
    this.minBandwidthBps = 50_000.0, // 50 kB/s
    this.minRttMs = 1.0,
  });

  /// Snapshot of the current view; copied so the caller can't mutate
  /// internal state.
  Map<String, BondedLinkState> get states {
    return Map<String, BondedLinkState>.unmodifiable(_states);
  }

  int get linkCount => _states.length;

  /// Replace the link set. Existing inflight counters for links that survive
  /// the update are preserved; entries that disappear have their inflight
  /// silently dropped (the session driver should call [completeSend] before
  /// removing a link to keep accounting clean).
  void updateLinks(Iterable<BondedLinkState> incoming) {
    Map<String, BondedLinkState> nextStates = <String, BondedLinkState>{};
    for (BondedLinkState s in incoming) {
      BondedLinkState? prior = _states[s.linkId];
      nextStates[s.linkId] = s.copyWith(
        inflightBytes: prior?.inflightBytes ?? s.inflightBytes,
      );
    }
    _states
      ..clear()
      ..addAll(nextStates);
    _order
      ..clear()
      ..addAll(nextStates.keys);
    if (_order.isEmpty) {
      _rrCursor = 0;
    } else {
      _rrCursor %= _order.length;
    }
  }

  /// Pick a link for an outbound packet of [bytes] bytes. Returns null when
  /// no link is eligible — callers should treat this as "drop or queue".
  /// The scheduler does not block / queue itself.
  ///
  /// [inflightFraction] scales the BDP cap used in the credit formula.
  /// Speed mode passes 1.0 by default. Values outside `[0.05, 1.0]`
  /// are clamped to keep the math sensible.
  BondedSchedulingDecision? pickLink({
    required int bytes,
    double inflightFraction = 1.0,
  }) {
    if (_states.isEmpty || bytes <= 0) {
      return null;
    }
    double frac = inflightFraction;
    if (frac < 0.05) frac = 0.05;
    if (frac > 1.0) frac = 1.0;
    List<String> eligible = _eligibleByPriority();
    if (eligible.isEmpty) {
      return null;
    }

    double bestCredit = double.negativeInfinity;
    String? winner;
    for (String id in eligible) {
      BondedLinkState s = _states[id]!;
      double credit = _credit(s, frac);
      if (credit > bestCredit) {
        bestCredit = credit;
        winner = id;
      }
    }

    bool wasRR = false;
    if (winner == null) {
      // Should be unreachable because eligible was non-empty, but guard for
      // arithmetic surprises (-inf vs -inf etc.).
      winner = eligible.first;
      wasRR = true;
      bestCredit = 0.0;
    } else {
      // If two or more eligible links share `bestCredit`, prefer the RR
      // fallback so weights converge to the configured ratio.
      int tieCount = 0;
      for (String id in eligible) {
        if ((bestCredit - _credit(_states[id]!, frac)).abs() < 1e-6) {
          tieCount++;
        }
      }
      if (tieCount > 1) {
        winner = _roundRobinPick(eligible);
        wasRR = true;
      }
    }

    // Book the bytes as inflight so subsequent picks see realistic credit.
    BondedLinkState s = _states[winner]!;
    _states[winner] = s.copyWith(inflightBytes: s.inflightBytes + bytes);
    return BondedSchedulingDecision(
      linkId: winner,
      wireId: s.wireId,
      credit: bestCredit,
      wasRoundRobinFallback: wasRR,
    );
  }

  /// Manually book [bytes] of inflight on [linkId]. Strategies that
  /// synthesise their picks (e.g. Redundant duplicating to every link)
  /// call this directly so the scheduler's credit math stays accurate
  /// when the user later switches back to Speed mode.
  void bookInflight(String linkId, int bytes) {
    BondedLinkState? s = _states[linkId];
    if (s == null) return;
    _states[linkId] = s.copyWith(inflightBytes: s.inflightBytes + bytes);
  }

  /// Release [bytes] from the link's inflight counter, signalling that the
  /// packet has been ACKed (or given up on after a timeout — either way the
  /// scheduler should stop pretending those bytes are still in flight).
  void completeSend(String linkId, int bytes) {
    BondedLinkState? s = _states[linkId];
    if (s == null) {
      return;
    }
    int next = math.max(0, s.inflightBytes - bytes);
    _states[linkId] = s.copyWith(inflightBytes: next);
  }

  /// Public credit accessor. Strategies consult it when they want to
  /// consider non-default link populations without
  /// invoking the booking side-effect of [pickLink]. Returns 0 when the
  /// link is unknown.
  double creditFor(String linkId, {double inflightFraction = 1.0}) {
    BondedLinkState? s = _states[linkId];
    if (s == null) {
      return 0.0;
    }
    double frac = inflightFraction.clamp(0.05, 1.0).toDouble();
    return _credit(s, frac);
  }

  /// Diagnostic alias retained for tests written before [creditFor] was
  /// promoted to a stable API. New code should call [creditFor].
  double creditForTest(String linkId, {double inflightFraction = 1.0}) =>
      creditFor(linkId, inflightFraction: inflightFraction);

  /// Inflight bytes accessor used by tests + telemetry.
  int inflightForTest(String linkId) {
    return _states[linkId]?.inflightBytes ?? 0;
  }

  // ---------------------------------------------------------------------
  // internals
  // ---------------------------------------------------------------------

  double _credit(BondedLinkState s, double frac) {
    double bandwidth = s.bandwidthBps < minBandwidthBps
        ? minBandwidthBps
        : s.bandwidthBps;
    double rttSec = (s.rttMs < minRttMs ? minRttMs : s.rttMs) / 1000.0;
    // Bandwidth × delay = the BDP "slot" the link can absorb. Subtract
    // inflight to model remaining headroom. Healthier link → more credit.
    double bdp = bandwidth * rttSec * frac;
    return bdp - s.inflightBytes.toDouble();
  }

  /// Return the eligible links ordered by priority (primary → backup, never
  /// excluded). Unhealthy / disabled links are filtered. We do priority
  /// filtering as a separate pass so credit is only compared inside the
  /// "best" bucket the policy allows.
  List<String> _eligibleByPriority() {
    if (_states.isEmpty) {
      return const <String>[];
    }
    // Find the best bucket that has any healthy member.
    for (LinkPriority bucket in <LinkPriority>[
      LinkPriority.primary,
      LinkPriority.secondary,
      LinkPriority.backup,
    ]) {
      List<String> ids = <String>[];
      for (String id in _order) {
        BondedLinkState s = _states[id]!;
        if (s.priority != bucket) continue;
        if (s.status == LinkStatus.unhealthy ||
            s.status == LinkStatus.disabled) {
          continue;
        }
        ids.add(id);
      }
      if (ids.isNotEmpty) {
        return ids;
      }
    }
    return const <String>[];
  }

  String _roundRobinPick(List<String> eligible) {
    if (eligible.length == 1) {
      _rrCursor = (_rrCursor + 1) % _order.length;
      return eligible.first;
    }
    // Iterate the master order so weights stay deterministic — the eligible
    // list is already filtered by priority/health.
    int start = _rrCursor;
    String picked = eligible.first;
    for (int i = 0; i < _order.length; i++) {
      int idx = (start + i) % _order.length;
      String candidate = _order[idx];
      if (eligible.contains(candidate)) {
        picked = candidate;
        _rrCursor = (idx + 1) % _order.length;
        return picked;
      }
    }
    return picked;
  }
}
